source/ingress.go

```diff
@@ -61,31 +65,57 @@ func NewIngressSource(kubeClient kubernetes.Interface, namespace, annotationFilt
 		}
 	}
 
-	return &ingressSource{
+	// Use shared informer to listen for add/update/delete of ingresses in the specified namespace.
+	// Set resync period to 0, to prevent processing when nothing has changed.
+	informerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(kubeClient, 0, kubeinformers.WithNamespace(namespace))
+	ingressInformer := informerFactory.Extensions().V1beta1().Ingresses()
+
+	// Add default resource event handlers to properly initialize informer.
+	ingressInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
```

Comment:
Maybe I miss the point, but where is the UpdateFunc and DeleteFunc? Don't we support them too?

Reply:
I think this is just a dummy handler to initialize the informer
This PR does not act at all on events but just provides an informer based cache to avoid making list calls against the apiserver

Reply:
I was also checking the other PR with doesn't implement this. I was thinking it would act on events and you do something with it.

For instance if a service/ingress gets deleted or the annotation gets changed. Not only if new ingress/service gets created...

I would also assume that even it's not doing anything with those events, it dropping Updates and Deletions is not good.

Reply:
If it does nothing other than make less calls to the API, you can also change the interval of the controller.

Reply:
On large clusters listing services and pods is slow, having an informer means sync will be much faster. On some of our medium size clusters, generating all records from services and pods took 5-10mn. With this change it will be a few seconds.

And, yes the goal is also to add handlers to update the dns provider entries as soon as there is a change without waiting for the next sync

Reply:
> I think this is just a dummy handler to initialize the informer...

@lbernail is correct, I found that at least one (dummy) handler was required to be registered in order for the informer cache to initialize properly.

> I was also checking the other PR which doesn't implement this. I was thinking it would act on events and it does something with it.

That is what the other PR adds. And it [actually does](https://github.com/kubernetes-incubator/external-dns/pull/687/files#diff-4f68287bd6bb0af9aca42d78be711d9cR305) register `Controller.RunOnce` callback to `AddFunc`, `UpdateFunc`, and `DeleteFunc` for Services and Ingresses when the `--events` flag is true.

Since using the informer's _cache_ to retrieve resources is distinct from registering event handlers to actually do something when they are updated, I decided to split the PRs in case we only wanted the former, or we wanted to space them out a bit.  If we want to use the cache _and_ trigger the sync loop on adds/updates/deletes and we want both those changes to go in and once, we could merge the other PR and skip this one. The other PR is a superset of this change.

Reply:
Ok understood. Thanks for answering my questions. One more question in regards to the other PR running the `Controller.Runonce`. 

This will multiply the execution of calls for each event right? It's already running in a loop but will add the calls of execution on top of it.

Reply:
@njuettner : I think the best would be to use a `BoundedFrequencyRunner` (which is what we use in kube-proxy) to ensure that `Runonce` is run at least once every period and at most once every X seconds (when there a lot of updates in a short timeframe)

Reply:
To work well, this will probably require enabling caching for the cloud provider records (except for small setups)

Reply:
I agree, this make sense. Since this PR will not doing anything besides having less active API calls, I would merge this for now and we need to think how to implement the `BoundedFrequencyRunner` somehow.

What are your thoughts?

Reply:
Example from kube-proxy:
Create the runner (here we would replace `proxier.syncProxyRules` by `Runonce`, `syncPeriod` by the external-dns period we want and `minSyncPeriod` by something like ~5s for instance): https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L429

And then, invoke it on add/delete/updates by calling the BoundedFrequencyRunner Run method
https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L631

Reply:
Sounds good @njuettner, @lbernail. 

In the meantime,  the new `AddEventHandler` function in Service/Ingress in the other PR [uses a BoundedFrequencyRunner](https://github.com/kubernetes-incubator/external-dns/pull/687/files#diff-4f68287bd6bb0af9aca42d78be711d9cR298) to at least limit the _additional_ calls to `Controller.RunOnce` triggered by events.

I'm happy to work with @lbernail or anyone else to use `BoundedFrequencyRunner` in a way to control _all_ calls to `Controller.RunOnce`.

Reply:
@lbernail @jlamillan would good if you both could work on that 👍. It seems like this a really good idea. 
Thanks again for the fruitful conversation and your suggestions, highly appreciated.

In the meantime I'm happy to merge this PR.